### DISCUSSION TITLE
initQpStatefulProperties problem

### DISCUSSION THREAD

**Discussion by styyokuda** [idaholab/moose#24281] [Q&A General] [answered]
Hi,
I used a program like below a few years ago but it does now work with today's MOOSE.
I get the error message of
"The parameter 'initial_length' is being retrieved before being set."
Please help!
Sincerely, S. Thomas

For the C file:
template <>
InputParameters
validParams<InitialProb>()
{
  InputParameters params = validParams<Material>();
  params.addParam<Real>("initial_length",0.1, "initial length");
  return params;
}

InitialProb::InitialProb(const InputParameters & parameters)
  : Material(parameters),
  _initial_length(getParam<Real>("initial_length")),
  _value(declareProperty<Real>("value")),
  _length(declareProperty<Real>("length")),
  _length_old(getMaterialPropertyOld<Real>("length_old"))
{
}

void
InitialProb::initQpStatefulProperties()
{
  _length[_qp] = _initial_length;
}

void
InitialProb::computeQpProperties()
{
  float constant1 = 2, constant2 = 5;
  _length[_qp] = _length_old[_qp] + constant1;
  _value[_qp] = _length[_qp] * constant2;
}

For h file:
template <>
InputParameters validParams<InitialProb>();

class InitialProb : public Material
{
public:
  InitialProb(const InputParameters & parameters);

protected:
  virtual void initQpStatefulProperties() override;
  virtual void computeQpProperties() override;

private:
  Real _initial_length;
  MaterialProperty<Real> & _value;
  MaterialProperty<Real> & _length;
  const MaterialProperty<Real> & _length_old;
};

**Comment by hugary1995** [comment]
Move validParams() inside the class like so
```cpp
class InitialProb : public Material
{
public:
  static InputParameters validParams();

  // ...
};
```

In the source file change accordingly
```cpp
InputParameters
InitialProb::validParams()
{
  InputParameters params = validParams();
  params.addParam("initial_length", 0.1, "initial length");
  return params;
}
```

**Reply by GiudGiud** [reply, marked as answer]
see this newsletter for the announcement and steps to take around this deprecation 

https://mooseframework.inl.gov/newsletter/2020/2020_04.html#c1ba1ef6-13a6-4210-bedb-0b691fec2ce7

**Reply by styyokuda** [reply]
Thank you very much for your patience with me.
S. Thomas
